substrate/client/rpc-servers/src/utils.rs

```diff
@@ -0,0 +1,161 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Substrate RPC server utils.
+
+use std::{
+	error::Error as StdError,
+	net::{IpAddr, SocketAddr, ToSocketAddrs},
+	str::FromStr,
+};
+
+use hyper::{
+	header::{HeaderName, HeaderValue},
+	Request,
+};
+use jsonrpsee::{server::middleware::http::HostFilterLayer, RpcModule};
+use tower_http::cors::{AllowOrigin, CorsLayer};
+
+static X_FORWARDED_FOR: HeaderName = HeaderName::from_static("x-forwarded-for");
+
+/// Helper to read the ip address of the client `X_FORWARDED_FOR` header.
+pub(crate) fn read_ip_from_proxy<B>(req: &Request<B>) -> Option<IpAddr> {
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For
+	//
+	// "X-Forwarded-For" returns a list of ip addresses
+	//
+	// X-Forwarded-For: 203.0.113.195,198.51.100.178
+	if let Some(ips) = req.headers().get(&X_FORWARDED_FOR).and_then(|v| v.to_str().ok()) {
+		if let Some(proxy_ip) = ips.split_once(',').and_then(|(v, _)| IpAddr::from_str(v).ok()) {
+			// NOTE: we assume that ip addr is global
```

Comment:
checking if the ip address is global is a PITA, https://doc.rust-lang.org/std/net/struct.Ipv4Addr.html#method.is_global but shouldn't be an issue in practice. 

Enabling rate-limiting for a local rpc server doesn't make sense to me :)